internal/core/src/cachinglayer/lrucache/DList.cpp

```diff
@@ -0,0 +1,201 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+#include "cachinglayer/lrucache/DList.h"
+
+#include <mutex>
+#include <vector>
+
+#include <folly/futures/Future.h>
+#include <folly/futures/SharedPromise.h>
+
+#include "cachinglayer/Utils.h"
+#include "cachinglayer/lrucache/ListNode.h"
+#include "monitor/prometheus_client.h"
+
+namespace milvus::cachinglayer::internal {
+
+bool
+DList::reserveMemory(const ResourceUsage& size) {
+    std::unique_lock<std::mutex> list_lock(list_mtx_);
+    auto used = used_memory_.load();
+    if (max_memory_.CanHold(used + size)) {
+        used_memory_ += size;
+        return true;
+    }
+    if (tryEvict(used + size - max_memory_)) {
+        used_memory_ += size;
+        return true;
+    }
+    return false;
+}
+
+bool
+DList::tryEvict(const ResourceUsage& expected_eviction) {
+    std::vector<ListNode*> to_evict;
+    // items are evicted because they are not used for a while, thus it should be ok to lock them
+    // a little bit longer.
+    std::vector<std::unique_lock<std::shared_mutex>> item_locks;
+
+    ResourceUsage size_to_evict;
+
+    auto would_help = [&](const ResourceUsage& size) -> bool {
+        auto need_memory =
+            size_to_evict.memory_bytes < expected_eviction.memory_bytes;
+        auto need_disk =
+            size_to_evict.file_bytes < expected_eviction.file_bytes;
+        return (need_memory && size.memory_bytes > 0) ||
+               (need_disk && size.file_bytes > 0);
+    };
+
+    for (auto it = tail_; it != nullptr; it = it->next_) {
+        if (!would_help(it->size())) {
+            continue;
+        }
+        // use try_to_lock to avoid dead lock by failing immediately if the ListNode lock is already held.
+        auto& lock = item_locks.emplace_back(it->mtx_, std::try_to_lock);
+        // if lock failed, it means this ListNode will be used again, so we don't evict it anymore.
+        if (lock.owns_lock() && it->pin_count_ == 0) {
+            to_evict.push_back(it);
+            size_to_evict += it->size();
+            if (size_to_evict.CanHold(expected_eviction)) {
+                break;
+            }
+        } else {
+            // if we grabbed the lock only to find that the ListNode is pinned; or if we failed to lock
+            // the ListNode, we do not evict this ListNode.
+            item_locks.pop_back();
+        }
+    }
+    if (!size_to_evict.CanHold(expected_eviction)) {
+        return false;
+    }
+    for (auto* list_node : to_evict) {
+        auto size = list_node->size();
+        internal::cache_eviction_count(size.storage_type()).Increment();
+        popItem(list_node);
+        list_node->clear_data();
```

Comment:
could clear data throw exception? especially under disk io use case. there is a leak on the resources?

Reply:
clear_data() should not throw exception. Ideally the resource managing class (CellT) should handle any exceptions occurred during clean up, and that should be invisible to caching layer.

under the hood clear_data() directly sets the `std::unique_ptr<CellT> cell_` to `nullptr` to call the destructor, which should then clean up the resources. and destructors in c++ is not supposed to throw exception.

Reply:
if we want to have such a semantic: `"if clean_data() failed, we still mark the resource as used."`, what we can do is to force all `CellT` to have a `bool TryCleanUp` method, which should return false on fail, and we call `TryCleanUp` before setting it to nullptr. 

but this is not favorable:

1. if the exception happens half way in the clean up process, it is hard to recover the state.
2. in `tryToEvict` we have accumulated sufficient space, if we then allow skip-eviction, we may not have sufficient `expected_eviction`.